### ISSUE TITLE
Erroneous CS0266 on 'byte = byte & byte'

### ISSUE THREAD

**Issue by aehancoc** [dotnet/roslyn#74375]
**Version Used**:
- .NET 8.0.6
- .NET Core 3.1.32 (via https://dotnetfiddle.net/)
- .NET 4.7.2 (via https://dotnetfiddle.net/)

**Steps to Reproduce**:

[Observe this example.](https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoTCDMACZAlAdkwG9MBIZJKijPRpx4ATwBcBTPCPAXjwoA3JmbM2XPMH488AMh4iGTAL6YVQA=)

**Diagnostic Id**:

CS0266: Cannot implicitly convert type 'int' to 'byte'. An explicit conversion exists (are you missing a cast?)

**Expected Behavior**:

Because `System.Byte` implements `System.Numerics.IBitwiseOperators<byte,byte,byte>` [as seen here](https://learn.microsoft.com/en-us/dotnet/api/system.byte.system-numerics-ibitwiseoperators-system-byte-system-byte-system-byte--op_bitwiseand?view=net-8.0), this should compile without issue. This occurs even if using a constant, like `byte b = a & (byte)2;`.

Interestingly, this works fine in VB.NET ([example](https://sharplab.io/#v2:DYLgbgRgNAJiDUAfAkgWwA4HsBOAXAzgAQDKAnvrgKaoCwAUAJBMP0AKArhMAJYDGhAWUwx2wSoOGjKARnoMOXPiU6CAhtwB2ACgCUchgBFuqQqsIBBIgCFSVQgF5CsxoeOEIF67fGOz5jTCmcgCiAcoQ9KGBQiJiQA=)) and in F# ([example](https://sharplab.io/#v2:DYLgZgzgPsCmAuACAhiRAjAnvWiC8iAFFjgJQCMA3ALABQcS6aJuByiAZFypUA==)).

**Actual Behavior**:

Compilation error is raised, because for whatever reason the compiler elects to cast some part of this to an `int`. The easiest workaround (and what the VB.NET/F# examples decompile to) is `byte b = (byte)(a & a);`, but this feels clunky.

**Comment by aehancoc** [comment]
May be related to https://github.com/dotnet/roslyn/issues/42816

**Comment by hez2010** [comment]
`IBitwiseOperators<byte,byte,byte>` are implemented explicitly for `byte`, those methods are not public: https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Byte.cs,422

If you want to call into `IBitwiseOperators<byte,byte,byte>.op_And`, you need to go through type parameter:

```cs
T And<T>(T l, T r) where T : IBitwiseOperators<T, T, T> => l & r;

byte a = 1, b = 2;
c = And(a, b);
```

Note that exposing those methods public on `byte` is considered as a breaking change, as existing code expect it to be casted to `int`.

**Comment by aehancoc** [comment]
I suppose that makes sense. The default behaviour is just a bit confusing at first glance. I'm assuming that it stems from the CIL's internal representation of integers as being at least 32 bits? Kinda rubs against the entire concept of the byte/short types, I'd think.

I'm not expecting anything to change, given the backwards compatibility requirements, so unless you want to reply then feel free to close this. (That said, I'm curious why the default integer promotion rules wouldn't be adequate if the methods *were* to be made public, but I'm sure there's plenty of edge cases I'm not considering.)

**Comment by CyrusNajmabadi** [comment]
Consider: 

```c#
Foo(b1 & b2)
void Foo(byte b) { ... }
void Foo(int i) { ... }
```

This can't change meaning.
